lib/grpc/client/adapters/gun.ex

```diff
@@ -135,7 +135,42 @@ defmodule GRPC.Client.Adapters.Gun do
   end
 
   @impl true
-  def recv_headers(%{conn_pid: conn_pid}, %{stream_ref: stream_ref}, opts) do
+  def receive_data(
+        %{server_stream: true, channel: %{adapter_payload: adapter_payload}, payload: payload} =
+          stream,
+        opts
+      ) do
+    with {:ok, headers, is_fin} <- recv_headers(adapter_payload, payload, opts) do
+      case is_fin do
+        :fin -> []
+        :nofin -> response_stream(stream, opts)
+      end
+      |> then(&{:ok, &1})
+      |> maybe_return_headers(opts[:return_headers], %{headers: headers})
```

Comment:
I don't quite like `maybe_*` naming. I'd rather use plain `return_headers` or avoid the function altogether in favor of an in-place if

Reply:
It also doesn't just return headers because you use it to return a map which can also contain trailers